### PR TITLE
Codacy pylint

### PR DESCRIPTION
**Pull request by EmilyBourne**
Add a configuration file for pylint. Fix various codacy errors. Remove unused class `FCodegen`. Add brackets for bool cast in c to remove warnings

### Chosen pylint configuration:
All pylint errors are enabled by default. A selection of errors are disabled (see below)

### Currently disabled pylint errors:
**Too-many** : These errors can be parametrised instead of removing them to modify the max number allowed
- `too-many-arguments`, (max 5)
- `too-many-locals`, (max 15 variables in the namespace.)
- `too-many-instance-attributes`, (max 7)
- `too-many-statements`, (max 50)
- `too-many-branches`, (max 12)
- `too-many-public-methods`, (max 20)
- `too-many-nested-blocks`, (max 5)
- `too-many-lines`, (lines in module max 1000)
- `too-many-return-statements`, (max 6)
- `too-many-function-args`, (max ??, raised when we pass 3 positional arguments)
- `too-many-ancestors`, (max 7, Constant has 11)
- `too-few-public-methods`, (min 2)
- `line-too-long` (max 80 chars)

**no-else**
- `no-else-return`,
- `no-else-break`,
- `no-else-continue`,
- `no-else-raise`

**Other**
- `bad-whitespace`
- `invalid-name` (checks if naming convention is followed)
- `import-outside-toplevel`
- `fixme` (raised by TODO)
- `c-extension-no-member` (raised when using mpi)
- `no-name-in-module` (raised when using .pyh file)
- `import-error` (raised when using .pyh file)
- `superfluous-parens`,
- `useless-object-inheritance`,
- `trailing-newlines`,
- `bad-continuation`, (wrong indentation on multi-line object)
- `arguments-differ`, (Parameters differ from overridden method, raised by many _print_X functions existing in sympy)
- `misplaced-comparison-constant`, (e.g.: `pyccel/codegen/printing/fcode.py:328:18: C01233: Comparison should be str(getattr(i.source, 'name', i.source)) == 'mpi4py'`)
- `unused-argument`,
- `redefined-outer-name`, (raised when a local variable has the same name as something in `__main__`
- `not-callable`, (often raised when using sympy's function. If this is a problem then it should be revealed by tests)
- `assigning-non-slot`,
- `multiple-statements`, (raised by : `class MathIsclose (MathFunctionBool): pass` )
- `abstract-method`, (raised when a method contains only `raise NotImplementedError`)
- `redundant-keyword-arg`,
- `duplicate-code` (managed by codacy)
- `unused-wildcard-import` (raised by anything imported using `import *` that is not used)

### Currently generated pylint errors:
- `missing-class-docstring` (raised by 1 line classes such as math/numpy function classes)
- `missing-function-docstring`
- `missing-module-docstring`
- `wrong-import-order` (raised when standard imports are not first, etc)
- `unused-variable`
- `import-self` (a file should not import a module with the same name as it will probably not be found. Our code contains mpi.py and mpi4py.py)
- `undefined-variable`
- `unused-import`
- `dangerous-default-value` (used when the default value is modifiable (e.g. list) as the default is a static variable)
- `wildcard-import` (only raised for `import pyccel.errors.messages`)
- `protected-access` (Access to a protected member _symbol of a client class)
- `inconsistent-return-statements` (Either all return statements in a function should return an expression, or none of them should. This is usually raised when calling `errors.report` which removes the need for a return)
- `consider-using-in`
- `duplicate-key` (raised in pyccel/codegen/printing/luacode.py, most keys are "")
- `unexpected-keyword-arg`
- `no-value-for-parameter`
- `redefined-argument-from-local`
- `no-member`
- `empty-docstring`
- `super-init-not-called`
- `useless-super-delegation`
- `eval-used`
- `unnecessary-pass`
- `no-self-use` (Method could be a function, disabled in files in pyccel/codegen/printing/)
- `no-value-for-parameter`
- `signature-differs`
- `unsupported-assignment-operation`
- `simplifiable-if-statement`
- `useless-return`
- `cyclic-import`

### CODE REVIEW

**Comment by EmilyBourne** [comment]
To determine the errors to disable I ran pylint on the pyccel folder and looked at which errors I thought made sense. There may be some that I missed or that I disabled unnecessarily

**Comment by EmilyBourne** [comment]
It seems that the .pylintrc file needs to be in the master branch for us to use it

**Review by yguclu — CHANGES_REQUESTED** [review]
Thanks @EmilyBourne, this is an impressive cleanup of Pyccel!

Were all your changes to the code prompted by pylint errors and warnings?

**Comment by yguclu** [comment]
In the PR description I suggest you write first the pylint errors that are enabled, and then those that are disabled.

**Comment by EmilyBourne** [comment]
> Were all your changes to the code prompted by pylint errors and warnings?

Almost all, one or two are in response to the warnings currently raised by the c tests

**Comment by EmilyBourne** [comment]
> In the PR description I suggest you write first the pylint errors that are enabled, and then those that are disabled.

The list of pylint errors that are enabled is very long. They are all enabled by default, then some are removed. The list in the description is not a list of errors that are enabled, but merely a list of errors still flagged in our code

**Comment by yguclu** [comment]
> > In the PR description I suggest you write first the pylint errors that are enabled, and then those that are disabled.
> 
> The list of pylint errors that are enabled is very long. They are all enabled by default, then some are removed. The list in the description is not a list of errors that are enabled, but merely a list of errors still flagged in our code

I see, please make that clear in the description then!

File: pyccel/codegen/printing/luacode.py [hunk 1]
```diff
@@ -344,7 +345,7 @@ def _print_NaN(self, expr, _type=False):
         return "NAN"
 
     def _print_Piecewise(self, expr):
-        if expr.args[-1].cond != True:
+        if expr.args[-1].cond is not True:
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/luacode.py:348]: Pylint says to use `if not expr.args[-1].cond`

File: pyccel/codegen/printing/luacode.py [hunk 2]
```diff
@@ -357,7 +358,7 @@ def _print_Piecewise(self, expr):
         for i, (e, c) in enumerate(expr.args):
             if i == 0:
                 lines.append("if (%s) {" % self._print(c))
-            elif i == len(expr.args) - 1 and c == True:
+            elif i == len(expr.args) - 1 and c is True:
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/luacode.py:361]: Similarly to above pylint suggests using `elif i == len(expr.args) - 1 and c:` here. Do you find this more readable?

File: pyccel/codegen/printing/luacode.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/codegen/printing/luacode.py]: Fine with me!

File: pyccel/codegen/printing/luacode.py (hunk 2 above)
Comment (yguclu) [reply pyccel/codegen/printing/luacode.py:361]: Fine with me!

**Review by yguclu — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Review by saidctb — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Pull request merged**